pkg/backup/exec_hook.go

```diff
@@ -0,0 +1,127 @@
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/heptio/ark/pkg/apis/ark/v1"
+	remotecommandconsts "k8s.io/apimachinery/pkg/util/remotecommand"
+	kscheme "k8s.io/client-go/kubernetes/scheme"
+	kapiv1 "k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+// PodCommandExecutor is capable of using the Kubernetes exec protocol to execute a command in a
+// container in a pod.
+type PodCommandExecutor interface {
+	// Execute executes a command in a container in a pod. If input is specified, it is passed to the
+	// command as STDIN. If the command takes longer than the specified timeout, an error is returned.
+	Execute(podNamespace, podName string, hook *v1.ExecHook) (string, string, error)
+}
+
+type podCommandExecutor struct {
+	restClientConfig *rest.Config
+	restClient       rest.Interface
+}
+
+// NewPodCommandExecutor creates a new PodCommandExecutor.
+func NewPodCommandExecutor(restClientConfig *rest.Config, restClient rest.Interface) PodCommandExecutor {
+	return &podCommandExecutor{
+		restClientConfig: restClientConfig,
+		restClient:       restClient,
+	}
+}
+
+// Execute executes a command in a container in a pod. If input is specified, it is passed to the
+// command as STDIN. If the command takes longer than the specified timeout, an error is returned
+// (NOTE: it is not currently possible to ensure the command is terminated when the timeout occurs,
+// so it may continue to run in the background).
+func (e *podCommandExecutor) Execute(podNamespace, podName string, hook *v1.ExecHook) (string, string, error) {
+	if podNamespace == "" {
+		return "", "", errors.New("podNamespace is required")
+	}
+	if podName == "" {
+		return "", "", errors.New("podName is required")
+	}
+	if hook == nil {
+		return "", "", errors.New("hook is required")
+	}
+	if hook.Container == "" {
+		return "", "", errors.New("hook.container is required")
+	}
+	if len(hook.Command) == 0 {
+		return "", "", errors.New("hook.command is required")
+	}
+
+	req := e.restClient.Post().
+		Resource("pods").
+		Namespace(podNamespace).
+		Name(podName).
+		SubResource("exec")
+
+	req.VersionedParams(&kapiv1.PodExecOptions{
+		Container: hook.Container,
+		Command:   hook.Command,
+		Stdin:     hook.Input != "",
+		Stdout:    true,
+		Stderr:    true,
+	}, kscheme.ParameterCodec)
+
+	executor, err := remotecommand.NewExecutor(e.restClientConfig, "POST", req.URL())
+	if err != nil {
+		return "", "", err
+	}
+
+	var stdout, stderr bytes.Buffer
+
+	streamOptions := remotecommand.StreamOptions{
+		SupportedProtocols: remotecommandconsts.SupportedStreamingProtocols,
+		Stdout:             &stdout,
+		Stderr:             &stderr,
+	}
+	if hook.Input != "" {
+		streamOptions.Stdin = strings.NewReader(hook.Input)
+	}
+
+	errCh := make(chan error)
+
+	go func() {
+		err = executor.Stream(streamOptions)
+		errCh <- err
+	}()
+
+	var timeoutCh <-chan time.Time
+	if hook.Timeout.Duration > 0 {
```

Comment:
Do we want to treat this similarly to the annotation where we set a default timeout?  As I understand this code to work currently, if we omit the annotation, it's 30 seconds.  If we get here from the `Hooks` attribute we would instead wait indefinitely

Reply:
Yes, will do